### PR TITLE
[#IP-331]  Migrate fp-ts version from 1.x to 2.x

### PR DESCRIPTION
**Pull request by fabriziopapi**
Update fp-ts version from 1.x to 2.x

#### List of Changes
Removed italia-ts-commons dependencies.
Changed openapi files to remove italia-ts-commons references. 
Update fp-ts, io-ts, ts-commons and openapigen versions.
Fixed compilation errors due to different fp-ts 2.x version mode.
Fixed unit test compilation errors.

#### Motivation and Context
Reduce technical debt and enable dev to use new fp-ts feature.

#### How Has This Been Tested?

Already existing unit tests.

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by fabriziopapi** [comment]
PR in draft due to io-functions-cgn "IP_383_upgrade_to_fp_ts_2" branch references.

**Comment by balanza** [comment]
@fabriziopapi is it still draft?

**Comment by fabriziopapi** [comment]
> @fabriziopapi is it still draft?

The PR https://github.com/pagopa/io-functions-cgn/pull/53 is still in draft, so I still can not remove the branch reference (see comment above)

**Comment by michaeldisaro** [comment]
Unit test are failing @fabriziopapi, can you check them?

File: src/services/messagesService.ts [hunk 1]
```diff
@@ -183,39 +185,45 @@ export default class MessagesService {
...
-                    data: {
-                      ...legalMessageMetadata.cert_data.data,
-                      timestamp
+            pipe(
+              TE.fromEither(
+                pipe(
+                  StrictUTCISODateFromString.decode(
+                    legalMessageMetadata.cert_data.data.timestamp
+                  ),
+                  x => x,
```
Comment (michaeldisaro) [inline src/services/messagesService.ts:201]: to remove

File: src/services/redisSessionStorage.ts [hunk 2]
```diff
@@ -999,17 +1013,21 @@ export default class RedisSessionStorage extends RedisStorageUtils
...
   }
 
   private parseUser(value: string): Either<Error, User> {
-    return parseJSON<Error>(value, toError).chain(data =>
-      User.decode(data).mapLeft(
-        err => new Error(errorsToReadableMessages(err).join("/"))
+    return pipe(
+      E.parseJSON(value, E.toError),
+      E.chain(data =>
+        pipe(
```
Comment (michaeldisaro) [inline src/services/redisSessionStorage.ts:1027]: I think we can use:
```
flow(
  User.decode,
  ...
```

File: src/types/user.ts [hunk 3]
```diff
@@ -4,8 +4,8 @@
  */
 
 import * as express from "express";
-import { Either, isLeft, left, right } from "fp-ts/lib/Either";
-import { fromNullable, none, Option, some, tryCatch } from "fp-ts/lib/Option";
+import * as E from "fp-ts/lib/Either";
```
Comment (michaeldisaro) [inline src/types/user.ts:7]: I've seen sometimes fp-ts/xxx and sometime fp-ts/lib/xxx... can we take a choice?

File: src/types/user.ts [hunk 4]
```diff
@@ -14,6 +14,8 @@ import {
 import { IResponseErrorValidation } from "@pagopa/ts-commons/lib/responses";
 import { DOMParser } from "xmldom";
 
+import { Either } from "fp-ts/lib/Either";
```
Comment (michaeldisaro) [inline src/types/user.ts:17]: Either is imported on line 7 as E.Either.

File: src/utils/__tests__/date.test.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
-import { isNone, isSome } from "fp-ts/lib/Option";
+import * as E from "fp-ts/Either"
```
Comment (michaeldisaro) [inline src/utils/__tests__/date.test.ts:1]: ditto

File: src/utils/getErrorCodeFromResponse.ts [hunk 6]
```diff
@@ -1,6 +1,8 @@
 import { DOMParser } from "xmldom";
 
-import { none, Option, some, tryCatch } from "fp-ts/lib/Option";
+import * as O from "fp-ts/lib/Option";
+import { Option } from "fp-ts/lib/Option";
```
Comment (michaeldisaro) [inline src/utils/getErrorCodeFromResponse.ts:4]: This should be get from O.Option

File: src/utils/spid.ts [hunk 7]
```diff
@@ -12,6 +12,8 @@ import {
   PatternString
 } from "@pagopa/ts-commons/lib/strings";
 import { DOMParser } from "xmldom";
+import { Option } from "fp-ts/lib/Option";
```
Comment (michaeldisaro) [inline src/utils/spid.ts:15]: ditto

**Review by michaeldisaro — CHANGES_REQUESTED** [review]
I didn't focused on heavy refactoring, but on small simple and fast changes that could have been done while migrating.
It was a good chance to clean a bit the code.

File: src/utils/barcode.ts [hunk 8]
```diff
@@ -50,24 +47,27 @@ export function toBarcode(
...
-    svg: fromEither(toBufferSvg(options))
-  }).foldTaskEither(
-    errorOrString =>
-      fromEither(
-        left(
-          typeof errorOrString === "string"
-            ? new Error(errorOrString)
-            : errorOrString
+  return pipe(
+    AP.sequenceS(TE.ApplicativePar)({
```
Comment (gquadrati) [inline src/utils/barcode.ts:51]: To be sure, did you check that `sequenceS(taskEither)` was a parallel execution?

**Review by gquadrati — CHANGES_REQUESTED** [review]
Thank you, very useful work. I Just left some small comments

File: src/services/profileService.ts [hunk 9]
```diff
@@ -99,11 +101,14 @@ export default class ProfileService {
...
-          >(
-            _ => ResponseErrorInternal(errorsToReadableMessages(_).join(" / ")),
-            _ => ResponseSuccessJson(_)
+          return pipe(
+            response.value,
+            ExtendedProfileApi.decode,
+            E.mapLeft(_ =>
+              ResponseErrorInternal(errorsToReadableMessages(_).join(" / "))
+            ),
+            E.map(_ => ResponseSuccessJson(_)),
```
Comment (AleDore) [inline src/services/profileService.ts:110]:
```suggestion
            E.map(ResponseSuccessJson),
```

File: src/types/user.ts [hunk 10]
```diff
@@ -280,15 +289,21 @@ export const withUserFromRequest = async <T>(
...
-          )}`
+  pipe(
+    O.tryCatch(() => JSON.parse(from)),
+    O.fold(
+      () => E.left<string, unknown>(`Invalid JSON for User [${from}]`),
+      _ => E.right<string, unknown>(_)
+    ),
+    E.chain(json =>
+      pipe(
+        User.decode(json),
```
Comment (AleDore) [inline src/types/user.ts:300]:
```suggestion
      flow(
        User.decode,
```

File: src/utils/ognl.ts [hunk 11]
```diff
@@ -105,14 +101,18 @@ export const ognlTypeFor = <T>(
   prefix: string
 ): t.Type<T, T, unknown> =>
   new t.Type<T, T, unknown>(
-    "OGNL",
+    "KafkaProducerCompactConfigFromEnv",
```
Comment (AleDore) [inline src/utils/ognl.ts:104]: ?

File: src/services/redisSessionStorage.ts [hunk 12]
```diff
@@ -862,24 +870,23 @@ export default class RedisSessionStorage extends RedisStorageUtils
...
-                resolve(right<Error, User>(session));
-              }
-            );
-          },
+          flow(
+            E.mapLeft(error => resolve(E.left(error))),
+            E.map(session => {
+              resolve(E.right(session));
+            }),
+            E.toUnion
```
Comment (BurnedMarshal) [inline src/services/redisSessionStorage.ts:886]: this should be unnecessary

File: src/services/redisSessionStorage.ts [hunk 13]
```diff
@@ -953,9 +965,11 @@ export default class RedisSessionStorage extends RedisStorageUtils
                   ) &&
                   !(p.prefix === fimsTokenPrefix && p.value === user.fims_token)
               ),
-              (_1, { prefix, value }) => `${prefix}${value}`
+              R.collect(Ord)((_1, { prefix, value }) => `${prefix}${value}`)
```
Comment (BurnedMarshal) [inline src/services/redisSessionStorage.ts:968]: We are ordering the results here.
In other place we use the deprecated `R.collect` without the `Ord` constraint.
```suggestion
              R.collect((_1, { prefix, value }) => `${prefix}${value}`)
```
Maybe this is an improvement but the output is different.

**Review by BurnedMarshal — COMMENTED** [review]
LGTM on files:
- `redisSessionStorage`
- `redisStorageUtils`
- `redisUserMetadata`
- `authenticationController`
- `sessionController`
only some comments

File: src/services/redisSessionStorage.ts (hunk 12 above)
Comment (fabriziopapi) [reply src/services/redisSessionStorage.ts:886]: you are rigth, thanks!

File: src/services/redisSessionStorage.ts [hunk 14]
```diff
@@ -3,29 +3,20 @@
...
-  fromLeft,
-  TaskEither,
-  taskEither,
-  taskify,
-  tryCatch
-} from "fp-ts/lib/TaskEither";
+import * as A from "fp-ts/lib/Array";
+import * as E from "fp-ts/lib/Either";
+import * as O from "fp-ts/lib/Option";
+// import { collect, StrMap } from "fp-ts/lib/StrMap";
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/services/redisSessionStorage.ts:9]: you are right, thanks!

File: src/utils/ognl.ts (hunk 11 above)
Comment (fabriziopapi) [reply src/utils/ognl.ts:104]: it's a typo, thank you!

File: src/config.ts [hunk 15]
```diff
@@ -141,47 +152,59 @@ const SPID_TESTENV_URL = process.env.SPID_TESTENV_URL;
...
-  | undefined = fromNullable(process.env.STARTUP_IDPS_METADATA)
-  .map(_ =>
-    parseJSON(_, toError)
-      .chain<Record<string, string> | undefined>(_1 =>
-        STRINGS_RECORD.decode(_1).mapLeft(
-          err => new Error(errorsToReadableMessages(err).join(" / "))
+export const STARTUP_IDPS_METADATA: Record<string, string> | undefined = pipe(
+  process.env.STARTUP_IDPS_METADATA,
+  O.fromNullable,
+  O.map(_ =>
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/config.ts:158]: Too much _ to remove: let's do a dedicated PR 🥇

File: src/config.ts [hunk 16]
```diff
@@ -228,8 +257,8 @@ export const serviceProviderConfig: IServiceProviderConfig = {
   spidTestEnvUrl: SPID_TESTENV_URL,
   spidValidatorUrl: process.env.SPID_VALIDATOR_URL,
   strictResponseValidation: {
-    ...(isSome(maybeSpidTestenvOption) ? maybeSpidTestenvOption.value : {}),
-    ...(isSome(maybeSpidValidatorUrlOption)
+    ...(O.isSome(maybeSpidTestenvOption) ? maybeSpidTestenvOption.value : {}),
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/config.ts:260]: you are right, but I do not want to touch the code not related to migration. We could put this changes to another PR.

File: src/config.ts [hunk 17]
```diff
@@ -499,9 +547,13 @@ export const getClientProfileRedirectionUrl = (
   token: string
 ): UrlFromString => {
   const url = clientProfileRedirectionUrl.replace("{token}", token);
-  return UrlFromString.decode(url).getOrElseL(() => {
-    throw new Error("Invalid url");
-  });
+  return pipe(
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/config.ts:550]: you are right, thank you!

File: src/config.ts [hunk 18]
```diff
@@ -534,12 +586,17 @@ export const getClientErrorRedirectionUrl = (
...
   const errorParams = record
-    .collect(params, (key, value) => `${key}=${value}`)
+    .collect(S.Ord)((key, value) => `${key}=${value}`)(params)
     .join("&");
+
   const url = CLIENT_ERROR_REDIRECTION_URL.concat(`?${errorParams}`);
-  return UrlFromString.decode(url).getOrElseL(() => {
-    throw new Error("Invalid url");
-  });
+  return pipe(
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/config.ts:593]: you are right, thank you!

File: src/controllers/sessionLockController.ts [hunk 19]
```diff
@@ -42,39 +41,52 @@ export default class SessionLockController {
     | IResponseErrorValidation
     | IResponseSuccessJson<SuccessResponse>
   > =>
-    taskEither
-      .of<IResponseErrorInternal | IResponseErrorValidation, void>(void 0)
-      .chain(_ =>
-        fromEither(
-          FiscalCode.decode(req.params.fiscal_code).mapLeft(err =>
+    pipe(
+      TE.of(void 0),
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/controllers/sessionLockController.ts:45]: you are right, thank you!

File: src/controllers/sessionLockController.ts [hunk 20]
```diff
@@ -90,29 +102,31 @@ export default class SessionLockController {
     | IResponseErrorValidation
     | IResponseSuccessJson<SuccessResponse>
   > =>
-    taskEither
-      .of<IResponseErrorInternal | IResponseErrorValidation, void>(void 0)
+    pipe(
+      TE.of(void 0),
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/controllers/sessionLockController.ts:106]: 👍

File: src/controllers/ssoController.ts [hunk 21]
```diff
@@ -87,50 +82,39 @@ export const getUserForFIMS = (profileService: ProfileService) => (
...
+    pipe(
+      TE.tryCatch(
+        () => profileService.getProfile(user),
+        () => ResponseErrorInternal("Cannot retrieve profile")
+      ),
+      TE.chain(r =>
+        r.kind === "IResponseSuccessJson" ? TE.of(r.value) : TE.left(r)
+      ),
+      TE.chainW(userProfile =>
+        TE.fromEither(
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/controllers/ssoController.ts:94]: Sorry, I do not understand what to move at the endo of the pipe

File: src/controllers/zendeskController.ts [hunk 22]
```diff
@@ -69,59 +64,56 @@ export default class ZendeskController {
...
-                "Cannot create a valid Zendesk user from this profile"
+      pipe(
+        TE.tryCatch(
+          () => this.profileService.getProfile(user),
+          () => ResponseErrorInternal("Cannot retrieve profile")
+        ),
+        TE.chain(r =>
+          r.kind === "IResponseSuccessJson" ? TE.of(r) : TE.left(r)
+        ),
+        TE.map(r => r.value),
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/controllers/zendeskController.ts:75]: you are right, thank you!

File: src/controllers/__tests__/zendeskController.test.ts [hunk 23]
```diff
@@ -1,7 +1,6 @@
 /* tslint:disable:no-any */
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/controllers/__tests__/zendeskController.test.ts:1]: you are right, thank you!

File: src/services/messagesService.ts (hunk 1 above)
Comment (fabriziopapi) [reply src/services/messagesService.ts:201]: a typo 👍

File: src/services/redisSessionStorage.ts (hunk 2 above)
Comment (fabriziopapi) [reply src/services/redisSessionStorage.ts:1021]: you are right, thank you!

File: src/types/user.ts (hunk 3 above)
Comment (fabriziopapi) [reply src/types/user.ts:7]: removed useless lib. thank you!

File: src/types/user.ts (hunk 4 above)
Comment (fabriziopapi) [reply src/types/user.ts:17]: fixed. thnak you!

File: src/utils/__tests__/date.test.ts (hunk 5 above)
Comment (fabriziopapi) [reply src/utils/__tests__/date.test.ts:1]: Sorry, I do not understand the advise here

File: src/utils/getErrorCodeFromResponse.ts (hunk 6 above)
Comment (fabriziopapi) [reply src/utils/getErrorCodeFromResponse.ts:4]: fixed. thank you!

File: src/services/messagesService.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/services/messagesService.ts:201]: :urlodimuch:

File: src/utils/barcode.ts (hunk 8 above)
Comment (fabriziopapi) [reply src/utils/barcode.ts:51]: check done by the @BurnedMarshal and @gquadrati

File: src/controllers/authenticationController.ts [hunk 24]
```diff
@@ -389,17 +397,22 @@ export default class AuthenticationController {
...
+        UserIdentity.decode,
+        E.mapLeft(_ =>
+          ResponseErrorInternal("Unexpected User Identity data format.")
+        ),
+        E.map(_ =>
+          pipe(
+            _,
+            exactUserIdentityDecode,
+            E.mapLeft(_1 => ResponseErrorInternal("Exact decode failed.")),
+            E.map(_1 => ResponseSuccessJson<UserIdentity>(_1)),
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/controllers/authenticationController.ts:411]: we do not need a cast. fixed. thank you!

File: src/controllers/sessionController.ts [hunk 25]
```diff
@@ -82,16 +82,15 @@ export default class SessionController {
             ) as ZendeskToken)
       };
 
-      return (await this.sessionStorage.update(updatedUser)).fold<
-        IResponseErrorInternal | IResponseSuccessJson<PublicSession>
-      >(
-        err => {
+      return pipe(
+        await this.sessionStorage.update(updatedUser),
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/controllers/sessionController.ts:86]: you are right, but I do not want to change too much the code. We could perform this change in another PR.

File: src/services/redisSessionStorage.ts (hunk 12 above)
Comment (fabriziopapi) [reply src/services/redisSessionStorage.ts]: I had to revert this changes: the unit tests will stuck on it.

**Review by AleDore — APPROVED** [review]

**Pull request merged**